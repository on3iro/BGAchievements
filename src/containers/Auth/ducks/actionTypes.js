export const LOGIN_SUBMIT = 'LOGIN_SUBMIT'
export const LOGIN_REQUEST = 'LOGIN_REQUEST'
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS'
export const LOGIN_ERROR = 'LOGIN_ERROR'
export const LOGOUT = 'LOGOUT'
export const REGISTER_SUBMIT = 'REGISTER_SUBMIT'
export const REGISTER_REQUEST = 'REGISTER_REQUEST'
export const REGISTER_SUCCESS = 'REGISTER_SUCCESS'
export const REGISTER_ERROR = 'REGISTER_ERROR'
export const UPDATE_MAX_ACCOUNTS = 'UPDATE_MAX_ACCOUNTS'
