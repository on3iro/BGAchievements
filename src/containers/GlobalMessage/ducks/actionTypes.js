export const GLOBAL_ERROR = 'GLOBAL_ERROR'
export const GLOBAL_MESSAGE = 'GLOBAL_MESSAGE'
export const CLEAR_MESSAGE = 'CLEAR_MESSAGE'
