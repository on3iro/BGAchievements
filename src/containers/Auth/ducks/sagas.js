import axios from 'axios'
import { all, fork, call, put, takeLatest } from 'redux-saga/effects'
import { delay } from 'redux-saga'

import { BASE_URL } from 'config/constants'

import * as types from './actionTypes'
import * as globalMessageTypes from 'containers/GlobalMessage/ducks/actionTypes'

// Login
export function * requestLogin (action) {
  const url = `${BASE_URL}/login`
  const config = {
    headers: {'Content-Type': 'application/json'}
  }

  try {
    const response = yield call(axios.post, url, action.payload, config)

    yield put({ type: types.LOGIN_SUCCESS, payload: response.data })
  } catch (error) {
    yield put({ type: types.LOGIN_ERROR })
    // TODO remove hardcoded values
    // The server currently has trouble propagating error messages correctly
    // therefore we need to specify the most common error message, although
    // this might to lead to wrong messages (e.g. if there is indeed a NetworkError)
    const errorMessage = 'Email und/oder Passwort sind falsch!'
    yield put({ type: globalMessageTypes.GLOBAL_ERROR, payload: errorMessage })
    yield call(delay, 20000)
    yield put({ type: globalMessageTypes.CLEAR_MESSAGE })
  }
}

// Register
export function * requestRegister (action) {
  const url = `${BASE_URL}/users`
  const config = {
    headers: {'Content-Type': 'application/json'}
  }

  try {
    const response = yield call(axios.post, url, action.payload, config)
    yield put({ type: globalMessageTypes.GLOBAL_MESSAGE, payload: 'Registrierung erfolgreich!' })
    yield put({ type: types.REGISTER_SUCCESS, payload: response.data })
  } catch (error) {
    yield put({ type: types.REGISTER_ERROR, payload: error.message })
    // TODO remove hardcoded values
    // The server currently has trouble propagating error messages correctly
    // therefore we need to specify the most common error message, although
    // this might to lead to wrong messages (e.g. if there is indeed a NetworkError)
    const errorMessage = 'Ein Nutzer mit dieser Email ist bereits registriert!'
    yield put({ type: globalMessageTypes.GLOBAL_ERROR, payload: errorMessage })
    yield call(delay, 20000)
    yield put({ type: globalMessageTypes.CLEAR_MESSAGE })
  }
}

// Auth watcher
export function * handleAuth () {
  yield takeLatest(types.LOGIN_SUBMIT, requestLogin)
  yield takeLatest(types.REGISTER_SUBMIT, requestRegister)
}

export default function * () {
  yield all([
    fork(handleAuth)
  ])
}
