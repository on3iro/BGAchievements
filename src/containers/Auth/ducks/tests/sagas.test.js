import { takeLatest, call, put } from 'redux-saga/effects'
import axios from 'axios'

import * as sagas from '../sagas'
import * as types from '../actionTypes'

describe('handleAuthSaga', () => {
  const handleAuthSaga = sagas.handleAuth()

  it('should start task to watch LOGIN_SUBMIT', () => {
    const takeLatestDescriptor = handleAuthSaga.next().value
    const expectedYield = takeLatest(types.LOGIN_SUBMIT, sagas.requestLogin)
    expect(takeLatestDescriptor).toMatchSnapshot()
  })

  it('should start taks to watch REGISTER_SUBMIT', () => {
    const takeLatestDescriptor = handleAuthSaga.next().value
    const expectedYield = takeLatest(types.REGISTER_SUBMIT, sagas.requestRegister)
    expect(takeLatestDescriptor).toMatchSnapshot()
  })
})

describe('login', () => {
  describe('requestLogin generator', () => {
    const action = {
      payload: {
        test: 'test'
      }
    }
    let requestLoginGenerator

    beforeEach(() => {
      requestLoginGenerator = sagas.requestLogin(action)
    })

    it('should call axios.post', () => {
      const callDescriptor = requestLoginGenerator.next().value
      const expectedYield = call(axios.post, 'http://localhost:3030/api/auth/login', action.payload, { headers: {'Content-Type': 'application/json'}})
      expect(callDescriptor).toMatchSnapshot()
    })

    it('should put LOGIN_SUCCESS action with payload', () => {
      const response = {
        data: {
          id: 1,
          email: 'foo@bar.com',
          name: 'foo',
          token: 'abc123'
        }
      }
      requestLoginGenerator.next().value
      const putDescriptor = requestLoginGenerator.next(response).value
      const expectedYield = put({ type: types.LOGIN_SUCCESS,
        payload: {
          id: 1,
          email: 'foo@bar.com',
          name: 'foo',
          token: 'abc123'
        }})
      expect(putDescriptor).toMatchSnapshot()
    })

    it('should put LOGIN_ERROR action', () => {
      const response = new Error('Some error')
      requestLoginGenerator.next().value
      const putDescriptor = requestLoginGenerator.throw(response).value
      const expectedYield = put({ type: types.LOGIN_ERROR, payload: 'Some error' })
      expect(putDescriptor).toMatchSnapshot()
    })
  })
})

describe('register', () => {
  describe('requestRegister generator', () => {
    const action = {
      payload: {
        test: 'test'
      }
    }
    let requestRegisterGenerator

    beforeEach(() => {
      requestRegisterGenerator = sagas.requestRegister(action)
    })

    it('should call axios.post', () => {
      const callDescriptor = requestRegisterGenerator.next().value
      const expectedYield = call(axios.post, 'http://localhost:3030/api/auth/register', action.payload, { headers: {'Content-Type': 'application/json'}})
      expect(callDescriptor).toMatchSnapshot()
    })

    it('should put REGISTER_SUCCESS action with payload', () => {
      const response = {
        data: {
          id: 1,
          email: 'foo@bar.com',
          name: 'foo',
          token: 'abc123'
        }
      }
      requestRegisterGenerator.next().value
      const putDescriptor = requestRegisterGenerator.next(response).value
      const expectedYield = put({ type: types.REGISTER_SUCCESS,
        payload: {
          id: 1,
          email: 'foo@bar.com',
          name: 'foo',
          token: 'abc123'
        }})
      expect(putDescriptor).toMatchSnapshot()
    })

    it('should put REGISTER_ERROR action', () => {
      const response = new Error('Some error')
      requestRegisterGenerator.next().value
      const putDescriptor = requestRegisterGenerator.throw(response).value
      const expectedYield = put({ type: types.REGISTER_ERROR, payload: 'Some error' })
      expect(putDescriptor).toMatchSnapshot()
    })
  })
})
