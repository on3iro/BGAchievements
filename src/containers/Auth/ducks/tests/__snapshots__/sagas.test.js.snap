exports[`handleAuthSaga should start taks to watch REGISTER_SUBMIT 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "REGISTER_SUBMIT",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`

exports[`handleAuthSaga should start task to watch LOGIN_SUBMIT 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "LOGIN_SUBMIT",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`

exports[`login requestLogin generator should call axios.post 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://simplel.de:6392/login",
      Object {
        "test": "test",
      },
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`

exports[`login requestLogin generator should put LOGIN_ERROR action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "LOGIN_ERROR",
    },
    "channel": null,
  },
}
`

exports[`login requestLogin generator should put LOGIN_SUCCESS action with payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "email": "foo@bar.com",
        "id": 1,
        "name": "foo",
        "token": "abc123",
      },
      "type": "LOGIN_SUCCESS",
    },
    "channel": null,
  },
}
`

exports[`register requestRegister generator should call axios.post 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://simplel.de:6392/users",
      Object {
        "test": "test",
      },
      Object {
        "headers": Object {
          "Content-Type": "application/json",
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`

exports[`register requestRegister generator should put REGISTER_ERROR action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": "Some error",
      "type": "REGISTER_ERROR",
    },
    "channel": null,
  },
}
`

exports[`register requestRegister generator should put REGISTER_SUCCESS action with payload 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": "Registrierung erfolgreich!",
      "type": "GLOBAL_MESSAGE",
    },
    "channel": null,
  },
}
`
